design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,255 @@
+  - If there is a presence of volume policy(fs-backup/snapshot)  on the backup request that matches as an action for a volume we use the newer volume policy approach to get the list of the volumes for `fs-backup` action
+  - Else continue with the annotation based legacy approach workflow.
+
+- Modifications needed for `snapshot` action:
+  - In the [takePVSnapshot function](https://github.com/vmware-tanzu/velero/blob/d4128542590470b204a642ee43311921c11db880/pkg/backup/item_backupper.go#L508) we will check the PV fits the volume policy criteria and see if the associated action is `snapshot`
+  - If it is not snapshot then we skip the further workflow and avoid taking the snapshot of the PV
+  - Similarly, For csi snapshot of PVC object, we need to do similar changes in [executeAction() function](https://github.com/vmware-tanzu/velero/blob/512fe0dabdcb3bbf1ca68a9089056ae549663bcf/pkg/backup/item_backupper.go#L348). we will check the PVC fits the volume policy criteria and see if the associated action is `snapshot` via csi plugin
+  - If it is not snapshot then we skip the csi BIA execute action and avoid taking the snapshot of the PVC by not invoking the csi plugin action for the PVC
+
+**Note:** 
+- When we are using the `VolumePolicy` approach for backing up the volumes then the volume policy criteria and action need to be specific and explicit, there is no default behaviour, if a volume matches `fs-backup` action then `fs-backup` method will be used for that volume and if similarly if the volume matches the criteria for `snapshot` action then the snapshot workflow will be used for the volume backup.
+- Another thing to note is the workflow proposed in this design uses the legacy opt-in/opt-out approach as a fallback option. For instance, the user specifies a VolumePolicy but for a particular volume included in the backup there are no actions(fs-backup/snapshot) matching in the volume policy for that volume, in such a scenario the legacy approach will be used for backing up the particular volume.
+## Implementation
+
+- The implementation should be included in velero 1.14
+
+- We will introduce a `volumehelper` package under which the `volume_policy_helper.go` will consist of the functions that we will use through the backup workflow to accommodate volume policies for PVs and PVCs
+
+#### FS-Backup
+- Regarding `fs-backup` action to decide whether to use legacy annotation based approach or volume policy based approach:
+    - We will use the `volumehelper.ProcessVolumePolicyFSBackup` function from the `volumehelper` package
+    - Functions involved in processing `fs-backup` volume policy action will somewhat look like:
+
+```go
+func ProcessVolumePolicyFSBackup(pod *corev1api.Pod, volumePolicy *resourcepolicies.Policies, defaultVolumesToFsBackup bool, backupExcludePVC bool) ([]string, []string, error) {
```

Comment:
A suggestion, instead of passing `defaultVolumesToFsBackup` and `backupExcludePVC`, what about passing the backup Spec as a whole so the func won't change when more flags are added to the backup CR(I hope not)?

Reply:
extend-vol-policy-design